openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/LiS/man/man3/isastream.3.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) isastream.3.man,v 1.1.4.3 2006/04/05 09:42:34 brian Exp
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified 2006/04/05 09:42:34 by brian
.\"
.\" -------------------------------------------------------------------------
.\"
.\" isastream.3.man,v
.\" Revision 1.1.4.3  2006/04/05 09:42:34  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\" =========================================================================
.\"
.\" Copyright (C) 2000  John A. Boyd Jr.  protologos, LLC
.\"
.\" This library is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU Library General Public
.\" License as published by the Free Software Foundation; either
.\" version 2 of the License, or (at your option) any later version.
.\"
.\" This library is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" Library General Public License for more details.
.\"
.\" You should have received a copy of the GNU Library General Public
.\" License along with this library; if not, write to the
.\" Free Software Foundation, Inc., 59 Temple Place - Suite 330, Cambridge,
.\" MA 02139, USA.
.\"
.\" =========================================================================
.\"
.\"
.TH ISASTREAM 3 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Library Calls"
.SH NAME
.B isastream
\- determine whether an open file is \fISTREAMS\fR
.\"
.\"
.SH SYNOPSIS
.PP
.B "#include <stropts.h> "
.HP
.BI "int isastream( int " filedes " ); "
.\"
.\"
.SH ARGUMENTS
.PP
.TP
.I filedes
a file descriptor to test.
.\"
.\"
.SH DESCRIPTION
.PP
.BR isastream "() "
may be used to test whether or not an open file descriptor is for a
.I STREAMS
file (see
.IR STREAMS "(4)).  "
.\"
.\"
.SH "RETURN VALUE"
.PP
Upon successful completion,
.BR isastream "() "
returns 1 if the indicated file is a
.IR STREAM ,
and 0 if it is not.
.\"
.\"
.SH "ERROR HANDLING"
.PP
If
.I filedes
does not identify a valid open file,
.BR isastream "() "
should return -1 with
.I errno
set to
.RB [ EBADF ].
.\"
.\"
.SH "APPLICATION USAGE"
.PP
.BR isastream "() " 
is convenient under Linux for determining or at least estimating
a file's behavior, since (with Linux
.I STREAMS
being available) there are both
.I STREAMS
and
.RI "non-" STREAMS
implementations for, e.g.,
.I fifos
(also known as
.I "named pipes"
- see
.BR fifo "(4)) "
and
.I pipes
(see
.BR pipe "(3)).  "
.sp
.BR isastream "() "
is also useful in conjunction with
.BR fattach "(3) "
and
.BR fdetach "(3) "
to determine whether or not a
.I STREAMS
file has been attached to a path name.
.sp
Applications wishing to use the version of the
.BR isastream "() "
function described herein must link with the
.B LiS
.[
LiS
.]
library which provides it.  The Linux/GNU C library otherwise provides a stub
supporting calls to
.BR isastream "() "
which only generates an error indicating that it is not implemented.
.\"
.\"
.SH BUGS
.PP
.BR isastream ()
has no known bugs.
.\"
.\"
.SH EXAMPLES
.PP
None.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR pipe "(3), "
.BR fattach "(3), "
.BR fdetach "(3), "
.BR fattach "(8), "
.BR fdetach "(8), "
.BR fifo "(4), "
.BR STREAMS "(4) "
.\"
.\"
.SH COMPATIBILITY
.PP
.BR isastream ()
is compatible with SVID,
.[
svid
.]
XID,
.[
xbd5
.]
SUSv2,
.[
susv2
.]
SUSv3,
.[
susv3
.]
POSIX,
.IR Solaris \(rg,
.IR UnixWare \(rg
and LiS.
.[
LiS
.]
.\"with the following exceptions:
.\".IP \(em \w'\(em\(em'u
.\"
.\"
.SH "CONFORMING TO"
.PP
SVID,
.[
svid
.]
XID,
.[
xbd5
.]
SUSv2,
.[
susv2
.]
SUSv3,
.[
susv3
.]
POSIX.
.\"
.\"
.SH VERSIONS
.PP
This manpage was written for
.B LiS\c
.
.[
LiS
.]
.\"
.\"
.SH HISTORY
.PP
UNIX\(rg\ System V Release 4 (SVR4)
.[
svr4
.]
.\"
.\"
.SH AUTHOR
.PP
John Boyd, protologos LLC.
.[
$LIST$
.]
.T3


openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/streams/lib/libc/isastream.c
/*****************************************************************************

 @(#) isastream.c,v OpenSS7-0_9_2_D_rc2(0.9.2.8) 2005/12/08 00:59:58

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2005  OpenSS7 Corporation <http://www.openss7.com>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; either version 2 of the License, or (at your option) any later
 version.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 675 Mass
 Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified 2005/12/08 00:59:58 by brian

 *****************************************************************************/

#ident "@(#) isastream.c,v OpenSS7-0_9_2_D_rc2(0.9.2.8) 2005/12/08 00:59:58"

static char const ident[] =
    "isastream.c,v OpenSS7-0_9_2_D_rc2(0.9.2.8) 2005/12/08 00:59:58";

#define _XOPEN_SOURCE 600
#define _REENTRANT
#define _THREAD_SAFE

#include <sys/types.h>
#include <sys/ioctl.h>
#include <errno.h>
#include <stropts.h>

/**
 * @ingroup libLiS
 * @brief test a stream.
 * @param fd a file descriptor to test.
 *
 * isastream() cannot contain a thread cancellation point (SUS/XOPEN/POSIX).
 * Because isastream() consists of a single system call, asynchronous thread
 * cancellation protection is not required.
 */
int
isastream(int fd)
{
	if (ioctl(fd, I_ISASTREAM) == -1) {
		if (errno == EBADF)
			return (-1);
		return (0);
	}
	return (1);
}
